File: backend/src/AnimeShop/API/Stock/ItemInstances.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Stock.ItemInstances (ItemInstancesAPI, itemInstancesServer) where

import AnimeShop.Prelude

type ItemInstancesAPI =
  "item_instances" :>
  (
    Get '[JSON] [Entity ItemInstance] :<|>
    ReqBody '[JSON] ItemInstance :> Post '[JSON] (Entity ItemInstance) :<|>
    ReqBody '[JSON] ItemInstanceId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity ItemInstance) :> Patch '[JSON] ()
  )

itemInstancesServer :: ServerT ItemInstancesAPI App
itemInstancesServer = getItemInstances :<|> postItemInstance :<|> deleteItemInstance :<|> patchItemInstance

getItemInstances :: App [Entity ItemInstance]
getItemInstances = runDb $ selectList [] []

postItemInstance :: ItemInstance -> App (Entity ItemInstance)
postItemInstance = runDb . insertEntity

deleteItemInstance :: ItemInstanceId -> App ()
deleteItemInstance = runDb . delete

patchItemInstance :: Entity ItemInstance -> App ()
patchItemInstance entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/stock/itemInstances.js
import { ADD_ITEM_INSTANCE, GET_ITEM_INSTANCES, DELETE_ITEM_INSTANCE } from '../../actions/types';

export function itemInstances(state = [], action) {
  switch (action.type) {
    case ADD_ITEM_INSTANCE:
      return [ ...state, action.itemInstance];
    case GET_ITEM_INSTANCES:
      return action.itemInstances;
    case DELETE_ITEM_INSTANCE:
      return state.filter(itemInstance => itemInstance.id !== action.id);
    default:
      return state;
  }
}
